### PR TITLE
[Dependencies][WIP] CocoaPods models

### PR DESCRIPTION
**Pull request by pepicrft**
**Note:** The PR is a work in progress so feel free to ignore it. I push it in this state to be able to pull from other environments. 

### Short description 📝
I'm adding the models to `TuistCore` and `ProjectDescription` to be able to declare Pod dependencies. After we merge this one, I'll work on the `pod install` and `pod update` commands. I'll leave the graph integration as the last piece.

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

File: Sources/ProjectDescription/Dependencies/Pod.swift [hunk 1]
```diff
@@ -0,0 +1,206 @@
+import Foundation
+
+/// A Pod dependency.
+///
+/// A Pod dependency can be either:
+///     - remote: A Git URL to the source of the package,
+///     and a requirement for the version of the package.
+///     - local: A relative path to the package.
+public enum Pod: Equatable, Codable {
+    case remote(url: String, requirement: Requirement)
```
Comment (pepicrft) [inline Sources/ProjectDescription/Dependencies/Pod.swift:10]: I think this one should be:

```swift
.local("Alamofire", path: "path/to/Alamofire")
.remote("Alamofire", requirement: .upToNextMajor("3.2.1"))
```

What do you think @danyf90 / @laxmorek ? Here's the [`Podfile`](https://guides.cocoapods.org/syntax/podfile.html) reference.

File: Sources/ProjectDescription/Dependencies/Pod.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            try container.encode(requirement, forKey: .requirement)
+        case let .local(path):
+            try container.encode(Kind.local, forKey: .kind)
+            try container.encode(path, forKey: .path)
+        }
+    }
+}
+
+extension Pod {
+    public enum Requirement: Codable, Equatable {
```
Comment (pepicrft) [inline Sources/ProjectDescription/Dependencies/Pod.swift:55]: If we end up having the same requirement as the Packages' I'll extract this into `DependencyRequirement` and share it across package and pod dependencies.

File: Sources/ProjectDescription/Dependencies/CocoapodsDependencies.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+/// Contains the description of dependencies that can be installed using CocoaPods.
+public struct CocoaPodsDependencies: Codable, Equatable {
```
Comment (pepicrft) [inline Sources/ProjectDescription/Dependencies/CocoapodsDependencies.swift:4]: @danyf90 / @laxmorek what do you think about this interface? Do you foresee the need for other attributes here?

Comment (laxmorek) [reply Sources/ProjectDescription/Dependencies/CocoapodsDependencies.swift:4]: Hey @pepicrft I think we can kick off this feature as MVP where list of pods may be enough. We will be extending it in the future.

File: Sources/ProjectDescription/Dependencies/Pod.swift (hunk 1 above)
Comment (laxmorek) [reply Sources/ProjectDescription/Dependencies/Pod.swift:10]: IMO enough to start with

File: Sources/ProjectDescription/Dependencies/Pod.swift (hunk 2 above)
Comment (laxmorek) [reply Sources/ProjectDescription/Dependencies/Pod.swift:55]: > and share it across package and pod dependencies

👍 

The extracting will be needed in the `TuistGraph` target too.

File: Sources/ProjectDescription/Dependencies/Pod.swift [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+        enum CodingKeys: String, CodingKey {
+            case kind
+            case revision
+            case branch
+            case minimumVersion
+            case maximumVersion
+        }
+
+        public init(from decoder: Decoder) throws {
```
Comment (laxmorek) [inline Sources/ProjectDescription/Dependencies/Pod.swift:71]: https://github.com/apple/swift-evolution/blob/main/proposals/0295-codable-synthesis-for-enums-with-associated-values.md 

It was implemented in Swift 5.5.

Could you remind me which minimum version of Swift we support? 🤔 Maybe we could consider bumping to 5.5.

**Review by laxmorek — APPROVED** [review]

**Comment by pepicrft** [comment]
Keep it open GitHub!

**Pull request closed without merging**